Source/JavaScriptCore/runtime/MicrotaskQueue.h

```diff
@@ -33,37 +33,65 @@
 
 namespace JSC {
 
+class JSGlobalObject;
+class MarkedMicrotaskDeque;
+class MicrotaskDispatcher;
 class MicrotaskQueue;
 class VM;
 
 class QueuedTask {
     WTF_MAKE_TZONE_ALLOCATED(QueuedTask);
     friend class MicrotaskQueue;
+    friend class MarkedMicrotaskDeque;
 public:
     static constexpr unsigned maxArguments = 4;
 
-    QueuedTask(MicrotaskIdentifier identifier, JSValue job, JSValue argument0, JSValue argument1, JSValue argument2, JSValue argument3)
-        : m_identifier(identifier)
+    enum class Status : uint8_t {
+        Executed,
+        Stopped,
+        Suspended,
+    };
+
+    QueuedTask(RefPtr<MicrotaskDispatcher> dispatcher, MicrotaskIdentifier identifier, JSGlobalObject* globalObject, JSValue job, JSValue argument0, JSValue argument1, JSValue argument2, JSValue argument3)
+        : m_dispatcher(dispatcher)
+        , m_identifier(identifier)
+        , m_globalObject(globalObject)
         , m_job(job)
         , m_arguments { argument0, argument1, argument2, argument3 }
     {
     }
 
-    void run();
+    void setDispatcher(RefPtr<MicrotaskDispatcher>&& dispatcher)
+    {
+        m_dispatcher = WTFMove(dispatcher);
+    }
+
+    Status tryRun();
+    bool isRunnable();
 
     MicrotaskIdentifier identifier() const { return m_identifier; }
+    JSGlobalObject* globalObject() const { return m_globalObject; }
+    JSValue job() const { return m_job; }
+    std::span<const JSValue> arguments() const { return std::span { m_arguments }; }
 
 private:
+    RefPtr<MicrotaskDispatcher> m_dispatcher;
     MicrotaskIdentifier m_identifier;
-    JSValue m_job;
-    JSValue m_arguments[maxArguments];
+    JSGlobalObject* m_globalObject { nullptr };
```

Comment:
We should figure out a way to avoid having all these raw pointer member variables in JSC code at some point.